_posts/function-contracts.md

```diff
@@ -0,0 +1,405 @@
+    let result = {
+        // Preconditions
+        kani::assume(max != 0 && min != 0);
+
+        // Inlined first execution of `gcd`
+        if min > max {
+            std::mem::swap(&mut max, &mut min);
+        }
+
+        let rest = max % min;
+        if rest == 0 { min } else {
+            // Inlined recursion
+            let max = min;
+            let min = rest;
+                        // Make sure preconditions are respected for recursion
+                        assert!(max != 0 && min != 0);
+            let result = kani::any();
+            kani::assume(max % result == 0 && min % result == 0 && result != 0);
+            result
+        }
+    };
+    // Make sure postconditions hold
+    assert!(max % result == 0 && min % result == 0 && result != 0);
+}
+```
+
+First it is helpful to think about the the recursion as a sequence of individual steps that build on top of one another, like a pyramid. Each step performs the same computation, but with different inputs, because the previous step calls into it with `min` and `rest`, which differ from `max` and `min`. Now consider the first step. It is called with non-deterministic inputs that are only constrained by the precondition, e.g. the line `kani::assume(max != 0 && min != 0)`. Then a computation is performed to calculate `min` and `rest` that is used for the recursive call and at the end the postconditions are enforced with `assert!(max % result == 0 && ...)`. What does that mean? It means if this step passes verification we can be sure that the postconditions hold for the inputs we verified with, e.g. any combination of `max` and `min` that satisfies `max != 0 && min != 0`. Now lets consider again the recursive call. We call with `min` and `rest` which are not the same as `max` and `min`. However remember that actually what our first step verification proves not just that the postconditions hold for *a* `max` and `min`, but in fact for *any* `max` and `min`, so long as the precondition is satisfied. Therefore we can conclude that it will also hold for `gcd(min, rest)`, if `min` and `rest` satisfy the preconditions, e.g. `min != 0 && rest != 0`. And that is precisely what is enforced where the abstraction is employed with `assert!(max != 0 && min != 0)`.
+
+To put it another way, when we recurse, we are allowed to assume that the verification passed, because the step that we are currently verifying is being checked for any possible input, including the one we are recursing with. If a problem were to occur anywhere in the recursive calls, we would see the same problem as a verification failure of the first step.
+
+Notice that using this trick, we reduced the up to 64 unrollings to just one, a substantial win. What is even better that doing this requires no input from you. Kani does this automatically. Any function with a contract is always verified inductively. Side note: this works even if the recursive call is not direct but buried somewhere in the call chain.
```

Comment:
I meant for contracts.

Reply:
I can only think of one which is and isn't specific to inductive verification. Technically using induction requires that the return type implement `Arbitrary` because it want's to use the replacement. Normally a contract check does not require `Arbitrary`, only the replacement does. However there is currently no way to turn off generation of the replacement function, and so if the return type didn't implement `Arbitrary` it would fail to compile regardless of whether we use inductive verification or not.

Reply:
I don't think we need to add this to the text at this point! Thanks!